packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -118,22 +118,22 @@ export class PendingStateManager implements IDisposable {
 		return this.pendingMessagesCount !== 0;
 	}
 
-	public getLocalState(): IPendingLocalState | undefined {
+	public getLocalState(snapshotSequenceNumber?: number): IPendingLocalState {
 		assert(
 			this.initialMessages.isEmpty(),
 			0x2e9 /* "Must call getLocalState() after applying initial states" */,
 		);
-		if (!this.pendingMessages.isEmpty()) {
-			return {
-				pendingStates: [...this.savedOps, ...this.pendingMessages.toArray()].map(
-					(message) => {
-						// delete localOpMetadata since it may not be serializable
-						// and will be regenerated by applyStashedOp()
-						return { ...message, localOpMetadata: undefined };
-					},
-				),
-			};
-		}
+		return {
+			pendingStates: [...this.savedOps, ...this.pendingMessages.toArray()]
+				.filter(
+					(message) => message.referenceSequenceNumber >= (snapshotSequenceNumber ?? 0),
```

Comment:
we should only filter this.saveOps, and if pendingMessages has a referenceSequenceNumber < snapshotSequenceNumber it should throw an error